source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,193 @@
+}
+
+Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (insert_) {
+    ENVOY_STREAM_LOG(debug, "CacheFilter::encodeHeaders inserting body", *encoder_callbacks_);
+    // TODO(toddmgreer): Wait for the cache if necessary.
+    insert_->insertBody(
+        data, [](bool) {}, end_stream);
+  }
+  return Http::FilterDataStatus::Continue;
+}
+
+void CacheFilter::onHeaders(LookupResult&& result) {
+  if (!active()) {
+    return;
+  }
+  switch (result.cache_entry_status_) {
+  case CacheEntryStatus::RequiresValidation:
+  case CacheEntryStatus::FoundNotModified:
+  case CacheEntryStatus::UnsatisfiableRange:
+    NOT_IMPLEMENTED_GCOVR_EXCL_LINE; // We don't yet return or support these codes.
+  case CacheEntryStatus::Unusable:
+    decoder_callbacks_->continueDecoding();
+    return;
+  case CacheEntryStatus::Ok:
+    response_has_trailers_ = result.has_trailers_;
+    const bool end_stream = (result.content_length_ == 0 && !response_has_trailers_);
+    // TODO(toddmgreer): Calculate age per https://httpwg.org/specs/rfc7234.html#age.calculations
+    result.headers_->addReferenceKey(Http::Headers::get().Age, 0);
+    decoder_callbacks_->encodeHeaders(std::move(result.headers_), end_stream);
+    if (end_stream) {
+      return;
+    }
+    if (result.content_length_ > 0) {
+      remaining_body_.emplace_back(0, result.content_length_);
+      getBody();
+    } else {
+      lookup_->getTrailers(
+          [this](Http::ResponseTrailerMapPtr&& trailers) { onTrailersAsync(std::move(trailers)); });
+    }
+  }
+}
+
+void CacheFilter::onHeadersAsync(LookupResult&& result) {
+  post([this, status = result.cache_entry_status_, headers = result.headers_.release(),
+        response_ranges = std::move(result.response_ranges_),
+        content_length = result.content_length_, has_trailers = result.has_trailers_] {
+    onHeaders(LookupResult{status, absl::WrapUnique(headers), content_length, response_ranges,
+                           has_trailers});
+  });
+}
+
+void CacheFilter::getBody() {
+  ASSERT(!remaining_body_.empty(), "No reason to call getBody when there's no body to get.");
+  lookup_->getBody(remaining_body_[0],
+                   [this](Buffer::InstancePtr&& body) { onBody(std::move(body)); });
+}
+
+void CacheFilter::onBodyAsync(Buffer::InstancePtr&& body) {
+  post([this, body = body.release()] { onBody(absl::WrapUnique(body)); });
+}
+
+// TODO(toddmgreer): Handle downstream backpressure.
+void CacheFilter::onBody(Buffer::InstancePtr&& body) {
+  if (!active()) {
+    return;
+  }
+  if (remaining_body_.empty()) {
+    ASSERT(false, "CacheFilter doesn't call getBody unless there's more body to get, so this is a "
+                  "bogus callback.");
+    decoder_callbacks_->resetStream();
+    return;
+  }
+
+  if (!body) {
+    ASSERT(false, "Cache said it had a body, but isn't giving it to us.");
+    decoder_callbacks_->resetStream();
+    return;
+  }
```

Comment:
As it worth it to handle these cases right now? For simplicity can we just ASSERT or possibly RELEASE_ASSERT that the cache doesn't do something it isn't supposed to do?

Reply:
Done.